src/dma_mapping.rs

```diff
@@ -0,0 +1,77 @@
+// Copyright © 2019 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+//
+
+use crate::VfioContainer;
+use crate::{get_host_address_range, ExternalDmaMapping};
+use std::convert::TryInto;
+use std::io;
+use std::sync::Arc;
+use vm_memory::{GuestAddress, GuestAddressSpace, GuestMemoryAtomic, GuestMemoryMmap};
+
+/// This structure implements the ExternalDmaMapping trait. It is meant to
+/// be used when the caller tries to provide a way to update the mappings
+/// associated with a specific VFIO container.
+pub struct VfioDmaMapping {
+    container: Arc<VfioContainer>,
+    memory: GuestMemoryAtomic<GuestMemoryMmap>,
+}
+
+impl VfioDmaMapping {
+    /// Create a DmaMapping object.
+    ///
+    /// # Parameters
+    /// * `container`: VFIO container object.
+    /// * `memory·: guest memory to mmap.
+    pub fn new(container: Arc<VfioContainer>, memory: GuestMemoryAtomic<GuestMemoryMmap>) -> Self {
+        VfioDmaMapping { container, memory }
+    }
+}
+
+impl ExternalDmaMapping for VfioDmaMapping {
+    fn map(&self, iova: u64, gpa: u64, size: u64) -> std::result::Result<(), io::Error> {
+        let user_addr = if let Some(addr) = get_host_address_range(
+            &self.memory.memory(),
```

Comment:
This may fail to support memory hot-add, but we may treat it as a bug and fix it later.